Add `run download` command for fetching workflow artifacts

mislav opened the pull request:
Downloading artifacts per run: `gh run download 123`
- downloads all non-expired artifacts from a run
- optionally, specific artifacts can be selected using the `-p` flag

Downloading artifacts per repository: `gh run download -p name1 -p name2`
- downloads matching non-expired artifacts across all runs
- specifying the name pattern is required in this case

No arguments = error. Should we display some kind of selector in this case? If so, what should the options in the selector be?

Inspirations:
- [actions/download-artifact](https://github.com/actions/download-artifact)
- `gh release download`

TODO:
- [x] tests for downloading & zip extraction
- [x] when downloading multiple artifacts, organize extracted files in sub-directories based on artifact name
- [x] address [CodeQL warning](https://github.com/cli/cli/pull/3349/checks?check_run_id=2255987466)

Fixes #3321

mislav committed:
Add `run download` command for downloading workflow artifacts

mislav committed:
Add ARTIFACTS information to `run view`

mislav committed:
Restore Go < 1.16 compatibility

mislav committed:
Merge remote-tracking branch 'origin' into artifact-download

mislav committed:
Fix test after merge

vilmibm commented on pkg/cmd/run/download/download.go:
```diff
@@ -0,0 +1,129 @@
+		Args:  cobra.MaximumNArgs(1),
+		Example: heredoc.Doc(`
+		  # Download all artifacts generated by a workflow run
+		  $ gh run download <run-id>
```
all of the other `gh run` subcommands drop into the interactive gflow without arguments. if this is run with no arguments (ie no `--pattern` and no run id), do you think it's valid to drop into that prompt here, too?

mislav replied:
Yeah, I was wondering about what could the interactive flow be about! Do you think it should:
1. Display the list of recent runs;
2. Display the list of available per-repository artifacts?

vilmibm replied:
good question, sorry I missed you asking this in the PR body :facepalm: My thinking is something like:

```
$ gh run download
? Would you like to download artifacts for a specific run?
> No, let me choose repository level artifacts
   Run 1
   Another Run
? Which artifact would you like to download?
> Artifact 1
   Artifact 2
```

mislav replied:
Thanks! Since first choosing a run and then choosing an artifact _within_ that run would just present a subset of options that would otherwise be present when scoped across the whole repo, I went with just prompting with a multi-select containing last 10 non-expired artifacts in a repository.

Is multi-select even what we want, or do you think people would mostly just be interested in a single artifact in interactive mode? If so, we can change it to single-option select and rely on non-interactive mode for multi-archive download.

vilmibm replied:
I like this approach! I think it makes sense and I don't have a strong sense one way or the other which people would prefer so I'm cool with keeping it this way and see what the reception is.

vilmibm commented on pkg/cmd/run/download/download_test.go:
```diff
@@ -0,0 +1,181 @@
+			opts := &tt.opts
+			io, _, stdout, stderr := iostreams.Test()
+			opts.IO = io
+			platform := &stubPlatform{listResults: tt.artifacts}
```
I appreciate the code organization aspect of `opts.Platform` but completely stubbing it out leaves a sizable chunk of code uncovered by unit tests. I'd appreciate at least some `httpmock` based tests for the platform code even if the feature-oriented tests use a stubbed `Platform`.

mislav replied:
There are now HTTP-based tests in addition to the stubbed ones ✨

vilmibm commented on pkg/cmd/run/view/view.go:
```diff
@@ -5,13 +5,15 @@ import (
 	"github.com/cli/cli/api"
 	"github.com/cli/cli/internal/ghinstance"
 	"github.com/cli/cli/internal/ghrepo"
+	"github.com/cli/cli/pkg/cmd/run/download"
```
Should this dependency be done via a `shared` package instead of a direct subcommand dependency?

vilmibm commented on pkg/cmd/run/view/view_test.go:
```diff
@@ -146,6 +146,9 @@ func TestViewRun(t *testing.T) {
 				reg.Register(
 					httpmock.REST("GET", "repos/OWNER/REPO/actions/runs/3"),
 					httpmock.JSONResponse(shared.SuccessfulRun))
+				reg.Register(
```
seems worth testing the artifact printing in at least one case

vilmibm reviewed (commented):
I'm mostly happy with the usability and direction of this! I otherwise just have questions around testing and code organization.

mislav committed:
Have StringSet preserve original order of values

mislav committed:
Address `run download` feedback

- With no arguments in TTY mode, prompt which artifacts to download
- Change `--pattern` argument to be just `--name` and only do exact
  matching
- For multi-archive downloads, prefix the destination path with the name
  of the artifact
- Add tests exercising HTTP functionality
- Avoid "zipslip" path injection when extracting ZIP files
- Add tests for ZIP extraction

mislav committed:
Move `Artifact` to the "shared" package

mislav committed:
Add tests for artifact rendering in `run view`

mislav committed:
Merge remote-tracking branch 'origin' into artifact-download

vilmibm commented on pkg/cmd/run/download/download.go:
```diff
@@ -0,0 +1,187 @@
+		  $ gh run download <run-id> -n <name>
+
+		  # Download specific artifacts across all runs in a repository
+		  $ gh run download -n <name1> -n <name2>
```
If there is no argument behavior I usually list it in examples

vilmibm reviewed (approved):
Tiniest nitpick about Examples and it needs a fix for the Windows tests but otherwise this is great!

vilmibm committed:
make test windows friendly

vilmibm committed:
add an example
